### PR TITLE
Define `KOKKOS_VERSION_{MAJOR,MINOR,PATCH}` and `KOKKOS_VERSION_{LESS,GREATER,EQUAL}` macros

### PR DESCRIPTION
**Pull request by dalg24**
Add
* `KOKKOS_VERSION_MAJOR`
* `KOKKOS_VERSION_MINOR`
* `KOKKOS_VERSION_PATCH`
macros and draft ~~`KOKKOS_VERSION_COMPARE` so users can write things like~~
```C++
#if KOKKOS_VERSION_COMPARE(>=, 3, 7, 0)
```

**edit per review** `KOKKOS_VERSION_{LESS,GREATER,EQUAL,LESS_EQUAL,GREATER_EQUAL}`
```C++
#if KOKKOS_VERSION_GREATER_EQUAL(3, 7, 0)

### CODE REVIEW

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -61,6 +61,14 @@
  *  KOKKOS_ENABLE_CUDA_UVM            Use CUDA UVM for Cuda memory space.
  */
 
+#define KOKKOS_VERSION_COMPARE(COMPARE, MAJOR, MINOR, PATCH) \
+  (KOKKOS_VERSION COMPARE((MAJOR)*10000 + (MINOR)*100 + (PATCH)))
```
Comment (masterleinad) [inline core/src/Kokkos_Macros.hpp:65]: Somehow, I feel uncomfortable providing the comparison operator as an argument and would instead introduce `KOKKOS_VERSION_LTE` and `KOKKOS_VERSION_GTE` (or `KOKKOS_VERSION_LT` and `KOKKOS_VERSION_GT` but I think the equal part is used more).

**Review by masterleinad — COMMENTED** [review]
I think this is a good idea.

File: core/src/Kokkos_Macros.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/Kokkos_Macros.hpp:65]: I think I agree with Daniel on this one.

Comment (dalg24) [reply core/src/Kokkos_Macros.hpp]: Updated

**Review by masterleinad — APPROVED** [review]
Looks good to me.

**Review by crtrott — APPROVED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -132,6 +132,12 @@ set(Kokkos_VERSION_MINOR 7)
 set(Kokkos_VERSION_PATCH 99)
 set(Kokkos_VERSION "${Kokkos_VERSION_MAJOR}.${Kokkos_VERSION_MINOR}.${Kokkos_VERSION_PATCH}")
 math(EXPR KOKKOS_VERSION "${Kokkos_VERSION_MAJOR} * 10000 + ${Kokkos_VERSION_MINOR} * 100 + ${Kokkos_VERSION_PATCH}")
+# mathematical expressions below are not stricly necessary but they eliminate
+# the rather aggravating leading 0 in the releases patch version number, and,
+# in some way, are a sanity check for our arithmetic
+math(EXPR KOKKOS_VERSION_MAJOR "${KOKKOS_VERSION} / 10000")
+math(EXPR KOKKOS_VERSION_MINOR "${KOKKOS_VERSION} / 100 % 100")
+math(EXPR KOKKOS_VERSION_PATCH "${KOKKOS_VERSION} % 100")
```
Comment (dalg24) [inline CMakeLists.txt:140]: To be clear, the alternative is remove that block and do
```CMake
#cmakedefine KOKKOS_VERSION_MAJOR @Kokkos_VERSION_MAJOR@
 #cmakedefine KOKKOS_VERSION_MINOR @Kokkos_VERSION_MINOR@
 #cmakedefine KOKKOS_VERSION_PATCH @Kokkos_VERSION_PATCH@

**Review by fnrizzi — APPROVED** [review]

**Pull request merged**